--- a/packages/codegen/src/IRBuilder.cj
+++ b/packages/codegen/src/IRBuilder.cj
@@ -4614,7 +4614,8 @@
         srcStart: LLVMValueRef, dstStart: LLVMValueRef, copyLen: LLVMValueRef): LLVMValueRef {
         let elementSizeValue = GetSize_64(rawArrayTy.GetElementType())
         let copyLenI64 = CastIntegerToInt64(copyLen)
-        let dataSize = CreateBinary(ExprKind.MUL, copyLenI64, elementSizeValue, name: "arr.data.len")
+        let dataSize = CreateBinary(ExprKind.MUL, copyLenI64,
+            if (rawArrayTy.GetElementType().IsGeneric()) { GetInt64(8) } else { elementSizeValue }, name: "arr.data.len")
         let srcIndex = CreateBinary(ExprKind.MUL, CastIntegerToInt64(srcStart), elementSizeValue, name: "src.index")
         let dstIndex = CreateBinary(ExprKind.MUL, CastIntegerToInt64(dstStart), elementSizeValue, name: "dst.index")
         let arrayLayoutType = CGArrayType.GenerateArrayLayoutType(cgMod, rawArrayTy)
