diff --git a/packages/codegen/src/IntrinsicsDispatcher.cj b/packages/codegen/src/IntrinsicsDispatcher.cj
--- a/packages/codegen/src/IntrinsicsDispatcher.cj
+++ b/packages/codegen/src/IntrinsicsDispatcher.cj
@@ -739,7 +739,8 @@ private func CPointerWrite(irBuilder: IRBuilder2, args: ArrayList<Value>,
         let addr = irBuilder.CreateInBoundsGEP(unsafe { LLVMInt8TypeInContext(irBuilder.GetLLVMContext().raw) },
             pointerValue, indices, name: "ele.ptr")
+        let payloadPtr = irBuilder.GetPayloadFromObject(value.GetRawValue())
         InsertAsanInstrument(irBuilder, typeArgs, addr, "CJ_MCC_AsanWrite")
-        irBuilder.CallGCReadGenericPayload(addr, value.GetRawValue(), valueTySize)
+        irBuilder.CreateMemCpy(addr, 0u32, payloadPtr, 0u32, valueTySize)
         return LLVMNullValue()
     }
     let rawValue = value.GetRawValue()
